CRM/Case/XMLProcessor/Process.php

```diff
@@ -568,6 +572,106 @@ public function createActivity($activityTypeXML, &$params) {
+   *
+   * @param array $activityParams
+   * @param object $activityTypeXML
+   * @return int|null the ID of the default assignee contact or null if none.
+   */
+  protected function getDefaultAssigneeForActivity($activityParams, $activityTypeXML) {
+    if (!isset($activityTypeXML->default_assignee_type)) {
+      return NULL;
+    }
+
+    $this->setupDefaultAssigneeOptionIds();
+
+    switch($activityTypeXML->default_assignee_type) {
+      case $this->defaultAssigneeOptionsIds['BY_RELATIONSHIP']:
+        return $this->getDefaultAssigneeByRelationship($activityParams, $activityTypeXML);
+        break;
+      case $this->defaultAssigneeOptionsIds['SPECIFIC_CONTACT']:
+        return $this->getDefaultAssigneeBySpecificContact($activityTypeXML);
+        break;
+      case $this->defaultAssigneeOptionsIds['USER_CREATING_THE_CASE']:
+        return $activityParams['source_contact_id'];
+        break;
+      case $this->defaultAssigneeOptionsIds['NONE']:
+      default:
+        return null;
+    }
+  }
+
+  /**
+   * Fetches and caches the activity's default assignee options.
+   *
+   * @return array
+   */
+  protected function setupDefaultAssigneeOptionIds() {
+    if (!empty($this->defaultAssigneeOptionsIds)) {
+      return $defaultAssigneeOptionsIds;
+    }
+
+    $defaultAssigneeOptions =  civicrm_api3('OptionValue', 'get', array(
+      'option_group_id' => 'activity_default_assignee',
+      'options' => array('limit' => 0)
+    ));
+
+    foreach($defaultAssigneeOptions['values'] as $option) {
+      $this->defaultAssigneeOptionsIds[$option['name']] = $option['id'];
+    }
+  }
+
+  /**
+   * Returns the default assignee for the activity by searching for the target's
+   * contact relationship type defined in the activity's details.
+   *
+   * @param array $activityParams
+   * @param object $activityTypeXML
+   * @return int|null the ID of the default assignee contact or null if none.
+   */
+  protected function getDefaultAssigneeByRelationship($activityParams, $activityTypeXML) {
+    if (!isset($activityTypeXML->default_assignee_relationship)) {
+      return NULL;
+    }
+
+    $targetContactId = is_array($activityParams['target_contact_id'])
+      ? $activityParams['target_contact_id'][0]
+      : $activityParams['target_contact_id'];
+
+    $relationships = civicrm_api3('Relationship', 'get', array(
+      'contact_id_b' => $targetContactId,
+      'relationship_type_id.name' => $activityTypeXML->default_assignee_relationship,
+      'sequential' => 1
+    ));
+
+    if ($relationships['count']) {
+      return $relationships['values'][0]['contact_id_a'];
+    } else {
+      return NULL;
+    }
+  }
+
+  /**
+   * Returns the activity's default assignee for a specific contact if the contact exists,
+   * otherwise returns null.
+   *
+   * @param object $activityTypeXML
+   * @return int|null
+   */
+  protected function getDefaultAssigneeBySpecificContact($activityTypeXML) {
+    $contact = new CRM_Contact_BAO_Contact();
```

Comment:
Any reasons you're using the BAO here instead of the API? The Contacts can be soft deleted. If you use the API, I THINK, it will take care of only returning non-deleted contacts. The BAO will probably not care about that.

Reply:
This is why someone who knows should check noobs trying to make changes :man_facepalming: